Dynamic IIIF canvas dimensions

mspalti opened the pull request:
## References
* Fixes #8021

## Description
Because images have varying aspect ratios and sizes, our current static default dimensions result in suboptimal layouts in Mirador. This PR sets the default canvas dimensions based on the size of the first bitstream image. 

## Instructions for Reviewers
This approach uses the IIIF image server's metadata response to set the default canvas size. The image sever response conforms to the IIIF Image API version 2.1.1 (and is supported by Cantaloupe). Using the image server seems reliable and  easy to implement but it does make testing difficult. I have not included a new IT. Alternately, this could be done instead by checking the bitstream image headers in DSpace.

This solution improves the default image layout. But, it's still a good idea to add explicit height and width metadata to the bitstream and this is essential when a single Item contains images of varying sizes. 

List of changes in this PR:
* Canvas service now has a method that updates the fallback canvas dimensions
* A new utility method retrieves the dimensions from the image server via the info.json metadata.

**Include guidance for how to test or review your PR.** 
Reviewers should be able to create a new Item with "dspace.iiif.enabled" and add a single image bitstream. The canvas dimensions in the IIIF manifest (retrieved by Mirador) should exactly match the dimensions of the image.  

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mspalti committed:
Update default height and width values using the iiif image server.

mspalti committed:
Added javadocs.

mspalti committed:
Fixed IT.

mspalti commented:
The big advantage to relying on the image server for image dimensions is support for JPEG2000 files. ImageIO does not support jp2 natively and adding a plugin becomes an extra, non-trivial step. Any institution using jp2 files will already have an image server with support and the IIIF image api provides a reliable way to retrieve the image dimension. So I think the approach taken here is the correct one.

abollini reviewed (changes requested):
Thanks @mspalti I like the general idea here but I would recommend a different approach (it would not imply major code rework anyway).
The default width and height should be set to -1 to explicit request to dspace to discover the right image sizes. This can be done modifying the `org.dspace.app.iiif.service.utils.IIIFUtils.getCanvasWidth(Bitstream, Bundle, Item, int)` method and the corresponding Height method to call a new method that will decide between the default settings and some guess for the current bitstream/bundle/item.
In this way it is also possible to give more freedom in the configuration allowing to make the guess just using the first bitstream or the exact bitstream (more expensive).
This method must be cached by spring so that multiple request for the same bitstream will avoid to query each time the image server

abollini commented:
after having reviewed also https://github.com/DSpace/DSpace/pull/8074

I recomend to refactor this contribution reusing the `IIIFCanvasDimensionService` introduced here https://github.com/DSpace/DSpace/pull/8074/files#diff-afe60b9a926a2f57cb2cc64d227812c2a34f392840a7459135cf06c1dfa020f3

mspalti commented on dspace-iiif/src/main/java/org/dspace/app/iiif/service/AbstractResourceService.java:
```diff
@@ -43,8 +43,8 @@
     // Default canvas dimensions.
-    protected static final Integer DEFAULT_CANVAS_WIDTH_FALLBACK = 1200;
-    protected static final Integer DEFAULT_CANVAS_HEIGHT_FALLBACK = 1600;
+    protected static Integer defaultCanvasWidthFallback = 2200;
```
It's true that the portrait aspect ratio is more common since books and documents have lots of images. What I did was test to see if a portrait or a landscape ratio (like this one) got the best result overall. It seemed like a landscape ratio (roughly similar to the aspect ratio of the Mirador viewer window) gave better results. But I wouldn't place too much weight on that observation.

mspalti commented:
@abollini , I want to be sure I understand what you are suggesting. Are you proposing that:

- If default width and height are set to -1 in `iiif.cfg` then dspace will update the default dimensions for all images (expensive).
- If the default height and width are defined in iiif.cfg and are `> -1`, then that dimension will  always be used as the default.
- If the default dimension is NOT defined in iiif.cfg and the first bitstream has no `iiif.image.width` then the default dimension will be updated based on the dimensions of first image (less expensive).
- If the default dimension is NOT defined in iiif.cfg and the first image has dimension metadata then the original default dimensions (currently 2200 x 1600) will be used.

mspalti committed:
Update manifestService after rebase.

tdonohue commented:
@mspalti : I'm a little confused between all your PRs relating to Canvas Dimensions...I'm not sure I understand the "default" behavior you are going for myself.  If we are able to dynamically determine the Canvas Dimensions at runtime, then why would we need the CLI script added in #8074 ?  Which approach should a site choose?  Should they be encouraged to run the CLI script? Or should they use the dynamic approach and only manually modify dimensions if they dynamic approach doesn't work?

Currently, I'm a bit worried that this PR "conflicts" slightly with #8074....it seems like either we determine dimensions dynamically (this PR's approach), or we need a way to bulk add them automatically (as in #8074).  But, I'm not sure I understand why we'd need **both** this PR and 8074?  Am I missing something?

mspalti commented:
@tdonohue , good question. The original intent of this PR was to update the "default" canvas dimension that is used in the absence of height and width metadata. The new default would be calculated ONLY when the first bitstream in the ORIGINAL bundle lacks dimension metadata (e.g. a user uploads an image through the submission UI and sets `dspace.iiif.enabled`). That's a likely scenario. Of course, if the admin has the #8074 task running periodically, it's also temporary since metadata will eventually be added.

Our hard-coded "default" dimension is a guess and it's often going to be inaccurate. So the intent here is to improve that guess by basing it on the dimensions of the first image. The new guess will be imperfect in lots of cases but it will certainly be a more informed and often quite accurate. It's also an inexpensive operation since the guess requires only the dimensions of a single image even when an item contains many, many images (e.g. a book).

If I understand what @abollini  suggested, this PR might be extended to include a new configuration option for setting the dimensions of all canvases at runtime. (Maybe that's not quite what Andrea is suggesting.) This is certainly possible but yes it _is_ slightly in conflict with #8074 and IMHO the batch CLI approach is better for several reasons. I think it would be the recommended approach.

abollini commented:
hi, this PR has still a lot of value regardless to the #8074 one.
The #8074 is the best approach in terms of performance as it provides accurate information about all the bitstream computing these information only one time. Nevertheless, once an item has been created the IIIF viewer experience will be sub-optimal until the task is run.
The current PR assures that the item is immediately functional, including in the submission / workflow phase to allow an easier review.

Due to the performance penalties that imply looking to all the bitstreams to calculate the image size/canvas size I have recommended to
- set the configuration to `-1` to enable the new behavior. Any values different than that will rely only on the current metadata based solution
- guess must be accurate and calculated for each individual bitstream if a metadata based value is not provided
- eventually add a new configuration to accept the size of the first bitstream as a best guest

mspalti committed:
Merge branch 'main' into dynamic-default-dims

mspalti commented:
This sounds good, @abollini . The only thing I would recommend is that the default should be setting the dimension dynamically in the absence of metadata in the first image. This will typically be a better estimate than the hard-coded default (which is still needed).  

Seems like a useful optimization particularly for less experienced users.

The user overrides this default behavior by setting the configuration to -1 or by adding their own global default dimensions to configuration.

For -1 we still need a cache. I have some work on that stashed that I could quickly add into the project.  It would be nice if this is in 7.2 but not essential.

mspalti committed:
Shared image server query service

mspalti committed:
Bitstream cache.

mspalti committed:
Updated cache logging.

mspalti committed:
Added config for updating default dimension for all images w/o metadata.

mspalti commented:
@abollini @tdonohue : I've implemented the `-1 `configuration option and added a separate cache for bitstream dimensions. Both the dimensions and cache are working when I test. Unfortunately, because the dimensions are provided by the image server I haven't been able to write and IT for this.

The default is still to update the default dimension if the first image lacks metadata. I can change this if we agree that it's better to leave this out or make it a separate configuration option. At the moment I still think it makes sense as the default behavior.

tdonohue reviewed (changes requested):
@mspalti : I tried testing this today by simply enabling IIIF on an Item (ensuring it had no IIIF metadata on the image bitstream), and then setting:
```
iiif.canvas.default-width = -1
iiif.canvas.default-height = -1
```

Unfortunately, when I then view the Item page, the call to `/server/iiif/[uuid]/manifest` returns a 500 error code. The underlying error (in dspace.log) is a NullPointerException:
```
java.lang.NullPointerException
org.dspace.app.iiif.service.utils.IIIFUtils.setDefaultSize(IIIFUtils.java:431)
org.dspace.app.iiif.service.utils.IIIFUtils.getCanvasWidth(IIIFUtils.java:372)
```

It appears to be caused by a Connection Refused error in my logs...
```
2022-01-26 16:22:57,893 INFO  unknown unknown org.dspace.app.rest.cache.CacheLogger @ Canvas Dimension Cache Event Type:
 CREATED | Key: c7671d2f-c3a5-4d15-b570-27c7aedfe9b6
2022-01-26 16:22:57,906 ERROR unknown 7bb2fba7-4c8a-411e-bcd8-3502caef8a1d org.dspace.iiif.IIIFApiQueryServiceImpl @ Con
nection refused (Connection refused)
java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:?]
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399) ~[?:?]
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242) ~[?:?]
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224) ~[?:?]
        at java.net.Socket.connect(Socket.java:609) ~[?:?]
        at java.net.Socket.connect(Socket.java:558) ~[?:?]
        at sun.net.NetworkClient.doConnect(NetworkClient.java:182) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:474) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:569) ~[?:?]
        at sun.net.www.http.HttpClient.<init>(HttpClient.java:242) ~[?:?]
        at sun.net.www.http.HttpClient.New(HttpClient.java:341) ~[?:?]
        at sun.net.www.http.HttpClient.New(HttpClient.java:362) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1253) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:1015) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1592) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520) ~[?:?]
        at org.dspace.iiif.IIIFApiQueryServiceImpl.getImageDimensions(IIIFApiQueryServiceImpl.java:46) ~[dspace-api-7.2-
SNAPSHOT.jar:7.2-SNAPSHOT]
        at org.dspace.app.iiif.service.utils.IIIFUtils.getImageDimensions(IIIFUtils.java:314) ~[dspace-iiif-7.2-SNAPSHOT
.jar:7.2-SNAPSHOT]
        at org.dspace.app.iiif.service.utils.IIIFUtils.setDefaultSize(IIIFUtils.java:430) ~[dspace-iiif-7.2-SNAPSHOT.jar
:7.2-SNAPSHOT]
        at org.dspace.app.iiif.service.utils.IIIFUtils.getCanvasWidth(IIIFUtils.java:372) ~[dspace-iiif-7.2-SNAPSHOT.jar
:7.2-SNAPSHOT]
```

The strange thing is that everything on my end works fine if I comment out those two `iiif.canvas.default-*` configurations.  Mirador loads properly & I see the image served from Cantaloupe. 

Occasionally, I can get this PR to work...but I *always* get this 500 error when I set both configs to `-1` and then stop & restart both the Cantaloupe image server & the backend. Then the first view of the Item throws this 500 error.  Because it occasionally will work, it might be some sort of caching issue...but I'm honestly not sure yet.

One other very minor code comment inline below.

mspalti commented:
I haven't yet replicated the connection refused error. But I did discover that the cache is not being used when configured to `-1`. The reason is this:

https://stackoverflow.com/questions/12115996/spring-cache-cacheable-method-ignored-when-called-from-within-the-same-class

Working on a fix for the cache now. Perhaps that will solve the issue. Also updated how we handle the NPE. A manifest should still be returned in the event that a lookup fails for some reason.

mspalti committed:
Fixed dimension cache.

mspalti commented on dspace-iiif/src/main/java/org/dspace/app/iiif/service/AbstractResourceService.java:
```diff
@@ -55,23 +55,29 @@
     @Autowired
     ImageProfileUtil imageUtil;
 
+    ConfigurationService configurationService;
```
It's a pattern I saw somewhere else in our code. I assume the intent is to bind the subclasses and parent resource class more closely. But yeah, doing this only for ConfigurationService is a bit odd. I'll change it if there's time.

tdonohue replied:
@mspalti : Honestly, I wouldn't worry about it too much. It just jumped out at me as slightly odd...but it's not high priority to fix.

mspalti commented:
The canvasdimension cache is working again. Sorry I didn't notice that it was broken after the last batch of changes for the `-1` config.

As of now, if the image server lookup fails and we have a null response, I'm reverting to the existing fallback dimensions and logging an extra error message. A manifest is returned instead of 500.

@tdonohue , I'm not sure why the image server connection fails for you. One thought is that when you use the Mirador viewer image requests are coming from the Angular host and not the backend. That might be a clue.

tdonohue commented:
@mspalti : There may be something odd going on in my Docker environment with regards to the connection issues. I've realized I hit this same issue in testing two, unrelated PRs today.  I'll try again tomorrow to see if I have better luck.

mspalti committed:
Added REMOVED event to canvas cache logger config.

mspalti committed:
Merge branch 'main' into dynamic-default-dims

abollini reviewed (approved):
thanks @mspalti it looks good to me. I have just left a real minor things inline if you have the time to look to it but it is not blocking from my perspective

mspalti committed:
Updated iiif config.

tdonohue reviewed (approved):
👍 Thanks @mspalti !  I just retested this today, and it's working for me now.  Previously, I suspect my connection issues were Docker related.  Thanks for the hard work on this!

tdonohue commented:
@mspalti : One last request on this PR... could you please add Docs for the new configurations to the existing IIIF page? https://wiki.lyrasis.org/display/DSDOC7x/IIIF+Configuration

mspalti commented:
@tdonohue , sure, I will add to the current iiif configuration documentation.

There's actually quite a lot of undocumented iiif functionality right now. I think we need a new page  or pages to cover all of that and this feature might be something to cover in greater detail in the extended documentation.

tdonohue commented:
@mspalti : Feel free to create subpages under that main IIIF page if you wish.  If I feel to rename or move them, I'll do so after the fact.  But, you are more than welcome to simply create what pages make sense to you and let me know when you feel it's ready for "review", etc.
